processing/src/main/java/io/druid/segment/DimensionIndexer.java

```diff
@@ -257,37 +263,27 @@ DimensionSelector makeDimensionSelector(
   boolean ARRAY = false;
 
   /**
-   * Given a row value array from a TimeAndDims key, as described in the documentation for
+   * Given a row value array from a Row key, as described in the documentation for
    * compareUnsortedEncodedKeyComponents(), convert the unsorted encoded values to a list or array of actual values.
    *
    * If the key has one element, this method should return a single Object instead of an array or list, ignoring
    * the asList parameter.
    *
-   * @param key dimension value array from a TimeAndDims key
+   * @param key dimension value array from a Row key
    * @param asList if true, return an array; if false, return a list
    * @return single value, array, or list containing the actual values corresponding to the encoded values
    *         in the input array
    */
   Object convertUnsortedEncodedKeyComponentToActualArrayOrList(EncodedKeyComponentType key, boolean asList);
 
-
-  /**
-   * Given a row value array from a TimeAndDims key, as described in the documentation for
-   * compareUnsortedEncodedKeyComponents(), convert the unsorted encoded values to an array of sorted encoded values
-   * (i.e., sorted by their corresponding actual values)
-   *
-   * @param key dimension value array from a TimeAndDims key
-   * @return array containing the sorted encoded values corresponding to the unsorted encoded values in the input array
-   */
-  EncodedKeyComponentType convertUnsortedEncodedKeyComponentToSortedEncodedKeyComponent(EncodedKeyComponentType key);
-
+  ColumnValueSelector convertUnsortedValuesToSorted(ColumnValueSelector selectorWithUnsortedValues);
```

Comment:
@jon-wei could you please help to create a good javadoc for this method? I forgot why we need to sort dimension values before index merging.

Reply:
Or @gianm?

Reply:
I don't remember either. Definitely they must be sorted _after_ merging (dictionaries need to be sorted) but I forget if there is a specific reason why they need to be sorted before merging.

Reply:
Sorry, just saw this comment!

The `writeDimValueAndSetupDimConversion` step that builds the merged dictionary across indexes and the conversion buffers used for converting per-index dictionary IDs -> merged dictionary IDs builds the conversions from each index's sorted dictionary, so the Rowboat iterable returned by IncrementalIndexAdapter also needs to use the sorted ID space.

`DictionaryMergeIterator` uses a priority queue to merge the per-index dictionary values, so I guess that's why the sorted IDs are used now.

Reply:
Now I don't understand why merging is not broken, if it compares `IndexedInts` (in this PR; or `int[]`, before) just based on sorted values, without looking up the original values. E. g.
Index1:
null -> 0,
apple -> 1,
banana -> 2.

Index2:
null -> 0,
watermelon -> 1.

Then it will merge [watermelon] row from the second index before [banana] row from the first index, because it's sorted index is lower.

Reply:
`writeDimValueAndSetupDimConversion` would look at the actual String values to create an int buffer for each index to be merged, used as a mapping between sorted index-specific IDs -> sorted merged IDs.

Using that example, the dictionary building step would create merged dictionary:
null -> 0
apple -> 1
banana -> 2
watermelon -3

As part of that process, a conversion buffer for each index to be merged would be generated:
Index 1:
0 -> 0 (null)
1 -> 1 (apple)
2 -> 2 (banana)

Index 2:
0 -> 0 (null)
1 -> 3 (watermelon)

These conversions are applied by `MMappedIndexRowIterable` before the rows are fed into the row merging function

Reply:
Ok, thanks. I see now. Updated Javadoc comments.

However now as I clearly see the whole picture, I have another question... #5526